Disallow creating modules tagged as allowing unsafe-string when -force-safe-string is enabled

kit-ty-kate opened the pull request:
While testing every packages to see which ones are not compatible with force-safe-string I came across this error which looked different from any other errors I've seen:
```
#=== ERROR while compiling haxe.3.4.7 =========================================#
# context              2.0.3 | linux/x86_64 | ocaml-variants.4.07.1+force-safe-string | file:///home/opam/opam-repository
# path                 ~/.opam/4.07.1+force-safe-string/.opam-switch/build/haxe.3.4.7
# command              /usr/bin/env OCAMLPARAM=safe-string=0,_ make
# exit-code            2
# env-file             ~/.opam/log/haxe-1-abfa5b.env
# output-file          ~/.opam/log/haxe-1-abfa5b.out
### output ###
# make -C libs/extlib OCAMLOPT=ocamlopt OCAMLC=ocamlc native
# make[1]: Entering directory '/home/opam/.opam/4.07.1+force-safe-string/.opam-switch/build/haxe.3.4.7/libs/extlib'
# ocamlopt -g -a -o extLib.cmxa enum.mli bitSet.mli dynArray.mli extArray.mli extHashtbl.mli extList.mli extString.mli global.mli rbuffer.mli IO.mli option.mli pMap.mli std.mli uChar.mli uTF8.mli base64.mli unzip.mli refList.mli optParse.mli dllist.mli multiArray.mli sMap.mli enum.ml bitSet.ml dynArray.ml extArray.ml extHashtbl.ml extList.ml extString.ml global.ml rbuffer.ml IO.ml option.ml pMa[...]
# File "bitSet.mli", line 1:
# Error: Unit BitSet imports from Enum, compiled with -unsafe-string.
#        This compiler has been configured in strict safe-string mode (-force-safe-string)
# make[1]: *** [Makefile:26: extLib.cmxa] Error 2
# make[1]: Leaving directory '/home/opam/.opam/4.07.1+force-safe-string/.opam-switch/build/haxe.3.4.7/libs/extlib'
# make: *** [Makefile:103: libs] Error 2
```
Basically what happened is that unsafe-string is enabled by the use of `OCAMLPARAM` instead of the compiler argument `-unsafe-string`. The later raises an hard error when used with a compiler configured with `force-safe-string` but the former doesn't raise any error and even allows it to some degree by changing the value of `Clflags.unsafe_string` to `true`. Changing this value doesn't allow `bytes` to be equal to `string` when the compiler is configured with `force-safe-string` but it still compiles the module interface as if it was, and thus raises an error when compiling the next module (`BitSet`) because pulling module tagged with unsafe-string in a switch with `force-safe-string` enabled is forbidden.

Test case to reproduce the problem highlighted in the `haxe` build log above:
```
$ ocamlc -config | grep safe_string
safe_string: true
default_safe_string: true
$ cat test.ml 
let x = 1
$ cat main.ml 
let () = print_endline (string_of_int Test.x)
$ OCAMLPARAM=safe-string=0,_ ocamlopt -c test.ml
$ OCAMLPARAM=safe-string=0,_ ocamlopt -c main.ml
File "main.ml", line 1:
Error: Unit Main imports from Test, compiled with -unsafe-string.
       This compiler has been configured in strict safe-string mode (-force-safe-string)
```
In this PR I replace `Clflags.unsafe_string` by its getter/setter counterparts to check if setting this value is correct or not before it's actually set, and raise an hard error if not. I didn't know which exception to use to I just went with `Invalid_argument` but feel free to tell me what to use if this is not the right one to use.

kit-ty-kate committed:
Disallow creating modules tagged as allowing unsafe-string when -force-safe-string is enabled

dra27 commented on utils/clflags.ml:
```diff
@@ -167,10 +167,18 @@ let with_runtime = ref true;;         (* -with-runtime *)
+       invalid_arg "Internal error: OCaml has been configured with \
+                    -force-safe-string: unsafe-string is not available"
+     else
+       r := v)
```
FWIW, I think this would be clearer if the functions were separated based on `Config.safe_string` (it'll be easier to delete it later when the code is totally removed, as well):

```OCaml
let get_unsafe_string, set_unsafe_string = (* -safe-string / -unsafe-string *)
  if Config.safe_string then
    (fun () -> false), (function false -> () | true -> invalid_arg "...")
  else
    let r = ref (not Config.default_safe_string) in
    (fun () -> !r), (fun v -> r := v)
```

dra27 reviewed (changes requested):
If I read correctly, the `invalid_arg` is going to trigger if `OCAMLRUNPARAM` contains `safe-string=0` which is not right. It's correct for the other uses, since we expect `Main_args.mk_unsafe_string` to have prevented the function from ever being called.

So `Compenv.read_one_param` needs to raise a proper error (like the `Arg.Bad` in `Main_args.mk_unsafe_string`). I'm perhaps overly allergic to all the `ref`s, but instead of `List.iter (fun b -> b := f bool) options` in `setter`, this could just be `f bool` and then `clear` can contain the `List.iter` part, `clear_fun` is not needed, and the `"safe-string"` case can use `setter` directly with a custom function which either calls `Clflags.set_unsafe_string` or raises an appropriate error.

dra27 commented:
(this will ultimately want a `Changes` entry - it's a bug fix)

kit-ty-kate commented:
@dra27 I removed the raise of `Invalid_argument` and changed it to a warning instead. I think I addressed all your concerns (apart from the Changes entry). If there is anything else I should do, don't hesitate to tell me.

dra27 commented on utils/clflags.ml:
```diff
@@ -167,10 +169,18 @@ let with_runtime = ref true;;         (* -with-runtime *)
+    (function
+      | false -> ()
+      | true -> raise (ForbiddenValue
+                         "OCaml has been configured with -force-safe-string: \
```
The option is now called `--enable-force-safe-string`

dra27 commented on utils/clflags.mli:
```diff
@@ -189,7 +192,12 @@ val with_runtime : bool ref
+
+(** [set_unsafe_string true] will raise [ForbiddenValue] if force-safe-string
+    is enabled *)
+val set_unsafe_string : bool -> unit
```
If there's to be an `ocamldoc` comment, let's have it below and use `@raise` 🙂 

```OCaml
(** Controls the `unsafe_string` setting.
    @raise ForbiddenValue if passed [true] and force-safe-string is enabled *)
```

dra27 commented on driver/compenv.ml:
```diff
@@ -193,8 +196,9 @@ let check_bool ppf name s =
-  let clear name options s = setter ppf (fun b -> not b) name options s in
+  let set name options s =  setter ppf (fun r b -> r := b) name options s in
+  let clear name options s = setter ppf (fun r b -> r := not b) name options s in
+  let clear_fun name options s = setter ppf (fun f b -> f (not b)) name options s in
```
These two lines are more than 80 chars.

dra27 commented on utils/clflags.ml:
```diff
@@ -167,10 +169,18 @@ let with_runtime = ref true;;         (* -with-runtime *)
+      | false -> ()
+      | true -> raise (ForbiddenValue
+                         "OCaml has been configured with -force-safe-string: \
+                          unsafe-string is not available"))
```
It's slightly unfortunate that `Clflags` (where this option is morally called `-unsafe-string`, given that the module is supposed to be about the command-line flags!) now has the `OCAMLPARAM` name embedded, but this exception can only be raised when parsing that.

The alternative (if it bothers someone else enough) is to keep the `invalid_arg` you had before and do this test in `Compenv` and raise the exception there.

dra27 reviewed (commented):
Nits only - another dev should review this too.

kit-ty-kate commented:
Superseded by #10893
